packages/palette/src/elements/PhoneInput/PhoneInput.tsx

```diff
@@ -0,0 +1,501 @@
+        case "Enter":
+        case "Meta":
+        case "Shift":
+        case "Tab":
+          // Ignore
+          return
+
+        case "Escape":
+          event.preventDefault()
+          event.stopPropagation()
+
+          setDropdownVisible(false)
+          inputRef.current?.blur()
+          reset()
+
+          return
+      }
+    }
+
+    const handleCountryPickerKeydown = (
+      event: React.KeyboardEvent<HTMLDivElement>
+    ) => {
+      switch (event.key) {
+        case "Escape":
+          event.preventDefault()
+          event.stopPropagation()
+
+          setDropdownVisible(false)
+          countryPickerRef.current?.blur()
+          reset()
+
+          return
+
+        case "Enter":
+        case " ":
+          event.preventDefault()
+          event.stopPropagation()
+
+          if (!disabled) {
+            setDropdownVisible(true)
+          }
+          return
+      }
+    }
+
+    const handleSearchInputKeydown = (
+      event: React.KeyboardEvent<HTMLInputElement>
+    ) => {
+      switch (event.key) {
+        case "Tab":
+          if (event.shiftKey) {
+            // ignore
+          } else {
+            // move down to the list
+            event.preventDefault()
+            event.stopPropagation()
+            reset()
+
+            set({ cursor: 0, interactive: true })
+          }
+          return
+
+        default:
+          searchInputRef.current?.focus()
+      }
+    }
+
+    return (
+      <Box
+        onKeyDown={handleContainerKeydown}
+        ref={composeRefs(containerRef, containsFocusRef) as any}
+        width="100%"
+        className={className}
+        {...boxProps}
+      >
+        <ContainerBox
+          ref={anchorRef as any}
+          isDropdownVisible={isDropdownVisible}
+          hover={hover}
+          focus={focus}
+          error={error}
+          disabled={disabled}
+          placeholder={inputProps.placeholder}
+        >
+          <SelectContainer
+            ref={countryPickerRef as any}
+            disabled={disabled}
+            onClick={() => {
+              if (!disabled && !isDropdownVisible) {
+                setDropdownVisible(true)
+              } else {
+                setDropdownVisible(false)
+              }
+            }}
+            tabIndex={disabled ? -1 : 0}
+            onKeyDown={handleCountryPickerKeydown}
+          >
+            {selectedOption.text}
+          </SelectContainer>
+
+          <StyledInput
+            disabled={disabled}
+            ref={composeRefs(inputRef, forwardedRef) as any}
+            type="tel"
+            autoComplete="tel-national"
+            name={inputName}
+            maxLength={25}
+            placeholder={inputProps.placeholder}
+            onChange={(e) => {
+              const inputValue = e.target.value
+              const countryCodeValue = selectedOption.value
+              inputProps.onChange?.(e)
+
+              onInputChange?.({ inputValue, countryCodeValue })
```

Comment:
**suggestion:**

I am not sure if this makes a difference, but updating the internal state first might make the UI update faster (in the case `inputProps.onChange` takes a long time).

 ```suggestion
               onInputChange?.({ inputValue, countryCodeValue })

               inputProps.onChange?.(e)
```